rfd/0134-cloud-first-deployment.md

```diff
@@ -0,0 +1,203 @@
+ready for deployment by applying a label (e.g. `cloud/verified`) on the master
+PR corresponding to the change. The release manager then will approve the
+further rollout.
+
+Once the change is rolled out to production tenants, engineers will be expected
+to monitor error rates and Cloud metrics relevant to their changes.
+
+Let's now look at some of the edge case scenarios in the release process and
+how we'll be handling them.
+
+### Behavior-incompatible changes
+
+Teleport's component compatibility guarantees become somewhat obsolete in a
+Cloud-first world. Historically, we've been reserving behavior-incompatible
+changes for major releases. The concept of a major release, however, becomes
+non-existent in the Cloud-first world where any change, backwards compatible
+or not, gets deployed to all users within days.
+
+When applied to Cloud release model, the compatibility requirement becomes:
+
+"Any Cloud release must be always compatible with the previous Cloud release"
+
+Compatibility requirements affect control plane (proxy, auth and other services
+we host for our users), agents and clients:
+
+- Control plane. We have full control over it, deploy in tandem and make sure
+  these are always compatible with each other.
+- Agents deployed by users. We always keep backwards compatibilithy between
+  adjacent releases and rely on automatic upgrades to make sure they don't drift.
+- Clients. Clients require automatic upgrades to avoid breakage on behavior
+  incompatible changes. This is [in progress](https://github.com/gravitational/cloud/issues/4880).
+
+### Rollbacks
+
+Bad releases can happen even with the multi-staged rollout process. In that
+situation the process will be:
+
+- For the control plane, we will prefer rolling forward in most cases to a new
+  release with the addressed issue or a rolled back change. The control plane
+  release process described above will be fast enough to be able to push the
+  new release quickly.
+    - In a scenario where a major critical path is affected and a full "roll
+      forward" process is not fast enough, the deployment team will manually
+      perform release rollback using the same process that is used today.
+- For the agents, the rollback mechanism is supported via automatic upgrade
+  system's critical/version endpoints described in the [RFD109](https://github.com/gravitational/teleport/blob/master/rfd/0109-cloud-agent-upgrades.md).
+- For the clients, which don't support automatic upgrades yet, the details are
+  TBD as part of the [automatic upgrades](https://github.com/gravitational/cloud/issues/4880)
+  work but we will likely need to build some automatic downgrade mechanism in
+  case a client was auto-upgraded but a corresponding Github release was pulled.
+
+### Security releases
+
+Existing security release process is compatible with Cloud-first deployment.
+
+Both `-cloud.X` and regular self-hosted tags will be published from teleport-private
```

Comment:
Will a security release block subsequent "regular" releases if there is a disclosure/update window we need to honour before pushing the security fixes back to OSS teleport? This isn't a big deal now as releases are relatively infrequent, but may become an issue if we want to release daily/several times a week.

Reply:
Incidentally - is `teleport-private` expected to remain an identical-down-to-the-hash fork of OSS `teleport`?

Is it worth defining how we intend that relationship to work now that there is scope for `teleport` and `teleport-private` to diverge more quickly, thanks to more frequent releases and larger dev team? This RFD may not necessarily be the place for that, though.

Reply:
No changes are expected in how we handle security releases, or the private repo. We can still upgrade Cloud control plane (and users' agents, through auto-upgrades) however often we want, we would just delay porting the fixes to OSS in case of embargo like we're doing it now.

Reply:
Yes, that's how the current process works as well. While the embargo is in effect, if another release needs to be published it will have to be published off of teleport-private. In practice, we would usually freeze releasing until embargo lifts unless there's an urgent fix needs to be deployed.

Reply:
I guess I don't have a clear idea of two timeframes:
* How long we typically need to embargo a security release for
* How frequently we want to do cloud releases in this new model ("with Cloud releases being regularly cut from the main development branch").

If the former is somewhat longer than the latter, then security releases are going to impact the cloud release process described here which would be at odds with "Existing security release process is compatible with Cloud-first deployment."

Reply:
The embargo period is typically 1 week. We usually freeze public releases for that week but publish off of teleport-private if we need to (e.g. urgent fix). We will just do the same with Cloud-first.